services/cache2/lru_test.go

```diff
@@ -0,0 +1,505 @@
+	err = l.Get("200", &v)
+	require.Equal(t, err, ErrKeyNotFound, "should contain nothing")
+
+	err = l.Set("201", 301)
+	require.Nil(t, err)
+	err = l.Get("201", &v)
+	require.Nil(t, err)
+	require.Equal(t, 301, v)
+
+}
+
+func TestLRUExpire(t *testing.T) {
+	l := NewLRU(&LRUOptions{
+		Size:                   128,
+		DefaultExpiry:          1 * time.Second,
+		InvalidateClusterEvent: "",
+	})
+
+	l.SetWithDefaultExpiry("1", 1)
+	l.SetWithExpiry("3", 3, 0*time.Second)
+
+	time.Sleep(time.Second * 2)
+
+	var r1 int
+	err := l.Get("1", &r1)
+	require.Equal(t, err, ErrKeyNotFound, "should not exist")
+
+	var r2 int
+	err2 := l.Get("3", &r2)
+	require.Nil(t, err2, "should exist")
+	require.Equal(t, 3, r2)
+}
+
+func TestLRUMarshalUnMarshal(t *testing.T) {
+	l := NewLRU(&LRUOptions{
+		Size:                   1,
+		DefaultExpiry:          0,
+		InvalidateClusterEvent: "",
+	})
+
+	value1 := map[string]interface{}{
+		"key1": 1,
+		"key2": "value2",
+	}
+	err := l.Set("test", value1)
+
+	require.Nil(t, err)
+
+	var value2 map[string]interface{}
+	err = l.Get("test", &value2)
+	require.Nil(t, err)
+
+	v1, ok := value2["key1"].(int)
+	require.True(t, ok, "unable to cast value")
+	require.Equal(t, 1, v1)
+
+	v2, ok := value2["key2"].(string)
+	require.True(t, ok, "unable to cast value")
+	require.Equal(t, "value2", v2)
+
+	post := model.Post{
+		Id:            "id",
+		CreateAt:      11111,
+		UpdateAt:      11111,
+		DeleteAt:      11111,
+		EditAt:        111111,
+		IsPinned:      true,
+		UserId:        "UserId",
+		ChannelId:     "ChannelId",
+		RootId:        "RootId",
+		ParentId:      "ParentId",
+		OriginalId:    "OriginalId",
+		Message:       "OriginalId",
+		MessageSource: "MessageSource",
+		Type:          "Type",
+		Props: map[string]interface{}{
+			"key": "val",
+		},
+		Hashtags:      "Hashtags",
+		Filenames:     []string{"item1", "item2"},
+		FileIds:       []string{"item1", "item2"},
+		PendingPostId: "PendingPostId",
+		HasReactions:  true,
+		ReplyCount:    11111,
+		Metadata: &model.PostMetadata{
+			Embeds: []*model.PostEmbed{
+				{
+					Type: "Type",
+					URL:  "URL",
+					Data: "some data",
+				},
+				{
+					Type: "Type 2",
+					URL:  "URL 2",
+					Data: "some data 2",
+				},
+			},
+			Emojis: []*model.Emoji{
+				{
+					Id:   "id",
+					Name: "name",
+				},
+			},
+			Files: nil,
+			Images: map[string]*model.PostImage{
+				"key": {
+					Width:      1,
+					Height:     1,
+					Format:     "format",
+					FrameCount: 1,
+				},
+				"key2": {
+					Width:      999,
+					Height:     888,
+					Format:     "format 2",
+					FrameCount: 1000,
+				},
+			},
+			Reactions: []*model.Reaction{
+				{
+					UserId:    "user_id",
+					PostId:    "post_id",
+					EmojiName: "emoji_name",
+					CreateAt:  111,
+				},
+			},
+		},
+	}
+	err = l.Set("post", post)
+	require.Nil(t, err)
+
+	var p model.Post
+	err = l.Get("post", &p)
+	require.Nil(t, err)
+	require.True(t, reflect.DeepEqual(post, p))
```

Comment:
Wondering if we could use `require.Equal(t, post, p)` for this ? It uses `reflect.DeepEqual` underneath it.

Reply:
thanks I didn't know that :)